shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -73,8 +73,15 @@
       executorService = FlutterInjector.instance().executorService();
     }
 
-    public DartMessengerTaskQueue makeBackgroundTaskQueue() {
-      return new DefaultTaskQueue(executorService);
+    public DartMessengerTaskQueue makeBackgroundTaskQueue(TaskQueueOptions options) {
+      switch (options.getType()) {
+        case SERIAL:
+          return new SerialTaskQueue(executorService);
+        case CONCURRENT:
+          return new ConcurrentTaskQueue(executorService);
```

Comment:
The executor service itself might not be concurrent.

This seems a bit confusing. You might have given us a concurrent executor service and we force serial operations on it, but you might have given us a single threaded executor service and now we're not forcing serial operations on it but they'll still be serial. 

I wonder a little bit if we're abstracting away `ExecutorService` unnecessarily. At any rate, this should probably be called `ALLOW_CONCURRENCY` rather than `CONCURRENT`, and maybe call the other one `FORCE_SERIAL`.

Reply:
The API has to work for all platforms.  It's only a fluke of the ExecutorService API and the fact that it is injectable that you can end up in a situation where you could request a CONCURRENT TaskQueue but it is actually serial.  A developer would only get in this situation if they explicitly injected a serial ExecutorService.  This won't be a problem on iOS for example.

How about I add documentation to the FlutterInjector warning about passing in an ExecutorService that is single threaded?

Reply:
This is specific to Java right?  

We can document whatever but we know we're wrapping an executor that may or may not allow concurrency.  We shouldn't make consumers think it's concurrent when we don't really know.

Reply:
This is specific to Java, but there will be an equivalent for iOS.  I'd like the API's to match.  It will be unfortunate to call it "CONCURRENT" on iOS and "ALLOW_CONCURRENCY" on another platform.

The documentation can say something like: "CONCURRENT - allows the provided ExecutorService to run at full concurrency, meaning tasks can operate in parallel provided that the ExecutorService defined in FlutterInjector supports it."

For SERIAL, we don't have to say anything special since no matter what it will operate serially.

Reply:
Why wouldn't we just call it `ALLOW_CONCURRENCY` on both though?

I'm fine with `SERIAL` instead of `FORCE_SERIAL`, but I'd like to avoid developers thinking that asking for concurrency will mean it's actually concurrent. Even on iOS, I'm assuming that we'll want to allow some kind of injection around what creates new threads and whether a request for a new thread actually creates one.

Reply:
We won't need something like injecting the ExecutorService on iOS since it has a globally accessible threadpool (GCD).  If anything we would allow the TaskQueueFactory to be injected.  We could do that on Android too if we wanted but it would be a waste not to share the ExecutorService used for initialization.

The nomenclature used when discussing this will be "concurrent" not "allow concurrency" so by calling it concurrent it is matching how people will talk about it and search for it.  It's similar to the naming of [dispatch queues on iOS](https://developer.apple.com/library/archive/documentation/General/Conceptual/ConcurrencyProgrammingGuide/OperationQueues/OperationQueues.html).

I could make it share the ExecutorService only if it hasn't been set.  That would mean we can make sure we have an ExecutorService that has concurrency.  Then allow people to inject the TaskQueueFactory.  Do you like that better?

Reply:
What happens if you don't want this to be backed by GCD in iOS?

What happens if or when we back this by the engine's own concurrent workers, which we don't allow users to control the number of? 

Part of my point here is that I would rather not have people assume that asking for this will get them concurrent execution. If you're on a small device with limited resources, we might very intentionally make this not be concurrent even if the user asked for it to be concurrent so that we make sure UI tasks can progress.

Reply:
How about replacing `TaskQueueOptions.setType()` with a `TaskQueueOptions.setSerialized()` flag that defaults to true?

If you set the serialized flag, then the `TaskQueue` ensures that no concurrent calls will be made to the channel's message handler.  If it is not set, then the `TaskQueue` provides no guarantees and concurrency is determined by the underlying executor.

Reply:
> What happens if you don't want this to be backed by GCD in iOS?

Honestly, as someone that has done iOS development for 10 years, that would probably never happen.  There would be no benefit of not using the ubiquitous shared thread pool that is going to exist anyways. 

How about this: if you as for a concurrent TaskQueue and you get a task queue with one thread, there will be no logical errors possible.  In fact, it might be possible to say that an ExecutorService with one thread is both "concurrent" and "serial".  So worrying about that case isn't worthwhile since it cannot lead to logical errors and may be technically consistent.

Reply:
> How about replacing TaskQueueOptions.setType() with a TaskQueueOptions.setSerialized() flag that defaults to true?

I think I can live with that, do you like it @dnfield ?

Reply:
Works for me

Reply:
Okay, done.